### ISSUE TITLE
BWN networks

### ISSUE THREAD

**Issue by clydebailey** [AlexeyAB/darknet#1940]

**Comment by AlexeyAB** [comment]
@clydebailey Hi,

Just try to set `binary=1` instead of `xnor=1` in the `[convolutional]` layers in cfg-file.
But I didn't test it well.

**Comment by clydebailey** [comment]
@AlexeyAB 
Ok I will try that .

**Comment by AlexeyAB** [comment]
@ghimiredhikura Do you use this cfg-file? https://github.com/AlexeyAB/darknet/blob/master/cfg/tiny-yolo_xnor.cfg

**Comment by AlexeyAB** [comment]
@ghimiredhikura 

* How many iterations did you train?

* Also you can try to use this XNOR-cfg file (there are only 3 XNOR layers): [yolov2-tiny_obj_xnor_fp32.cfg.txt](https://github.com/AlexeyAB/darknet/files/2690280/yolov2-tiny_obj_xnor_fp32.cfg.txt)

**Comment by ghimiredhikura** [comment]
@AlexeyAB 

> * How many iterations did you train?

I did train more then 200000 iterations. 

> * Also you can try to use this XNOR-cfg file (there are only 3 XNOR layers): [yolov2-tiny_obj_xnor_fp32.cfg.txt](https://github.com/AlexeyAB/darknet/files/2690280/yolov2-tiny_obj_xnor_fp32.cfg.txt)

This way, I think mAP will be improved, but I am trying to train network similar to this one for hardware implementations. 
[A Lightweight YOLOv2: A Binarized CNN with A Parallel Support Vector Regression for an FPGA](http://delivery.acm.org/10.1145/3180000/3174266/p31-nakahara.pdf?ip=203.254.173.42&id=3174266&acc=ACTIVE%20SERVICE&key=4C7934339E2BD702%2E4C7934339E2BD702%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&__acm__=1545181117_98da60aa93e7e7726a76fa3d9aa424e8)
![image](https://user-images.githubusercontent.com/24696403/50192078-27c7b580-0374-11e9-84e8-5607311590cf.png)
Therefor I can't use only 3 layers as XNOR layers. 

Any suggestions please!

**Comment by AlexeyAB** [comment]
@ghimiredhikura 

If I understand correctly, they try to use network like this: [yolov2_xnor_fpga.cfg.txt](https://github.com/AlexeyAB/darknet/files/2694483/yolov2_xnor_fpga.cfg.txt) - try to train this model.

This is hybrid of yolov2.cfg and yolov2-tiny.cfg. With only 1st and two last float layers, other XNOR layers.

![image](https://user-images.githubusercontent.com/4096485/50215704-e3004680-0394-11e9-9276-15c652688f39.png)


Also check that you uncommented `difficult` in the `voc.data` file before doing `./darknet detector map`...:
```
classes= 20
train  = data/train_voc.txt
valid  = data/2007_test.txt
difficult = data/difficult_2007_test.txt
names = data/voc.names
backup = backup/
```

**Comment by ghimiredhikura** [comment]
@AlexeyAB,

Thank you for your quick response. Now I am training network that you suggested. I used pretrained weights from [darknet19 classifier](https://github.com/pjreddie/darknet/blob/master/cfg/extraction.cfg). 
Here is the training progress after >20000 iterations! 

![chart](https://user-images.githubusercontent.com/24696403/50263221-0dff9e80-0458-11e9-9086-db868bb8f193.jpg)

**Comment by AlexeyAB** [comment]
@ghimiredhikura 


Update your code from Github. And recompile.


----

Download this weights: http://pjreddie.com/media/files/yolov2-voc.weights

Do:
`./darknet partial cfg/yolov2-voc.cfg yolov2-voc.weights yolov2.conv.23 23`

And train using `yolov2.conv.23` pre-trained weights file.

**Comment by ghimiredhikura** [comment]
@AlexeyAB, 

Updated code from Github and run training over the weekend using yolov2.conv.23 extracted from http://pjreddie.com/media/files/yolov2-voc.weights. 

Here is training result and mAP status: 
![chart](https://user-images.githubusercontent.com/24696403/50388946-91760400-0766-11e9-81c9-402771be76ad.jpg)
Although in training graph, the mAP seems to be wrong, I calculated mAP using light_yolo repo (ofcourse using latest weights), but unfortunately it is still very low!
![image](https://user-images.githubusercontent.com/24696403/50388995-0d704c00-0767-11e9-9a0a-ca5e3d42b852.png)

**Comment by AlexeyAB** [comment]
@ghimiredhikura There is a bug in mAP calculation during Training if is used CUDNN=1. I will solve it.

**Comment by ghimiredhikura** [comment]
My gpu is 1080 ti. And training in Win 10 with visual studio.

**Comment by AlexeyAB** [comment]
@ghimiredhikura 

I added fix. I hope now mAP calculation during Training is fixed.
Update your code from GitHub.

----

>  I calculated mAP using light_yolo repo (ofcourse using latest weights), but unfortunately it is still very low!

I do not know why this is so.
Try to use this cfg-file, I modified it: [yolov2_xnor_fpga.cfg.txt](https://github.com/AlexeyAB/darknet/files/2707811/yolov2_xnor_fpga.cfg.txt)
And try to train it until mAP stops.

![image](https://user-images.githubusercontent.com/4096485/50407887-77cdce80-07f1-11e9-8206-32d69ba5ce38.png)

**Comment by ghimiredhikura** [comment]
@AlexeyAB 
I was training with updated code and new cfg file you attached above, but mAP is still very low! 
  
![chart](https://user-images.githubusercontent.com/24696403/50577825-11d9ea80-0e74-11e9-9d56-d1afb6f9c119.png)

**Comment by ghimiredhikura** [comment]
@AlexeyAB, Hi, 

Sorry for delay in getting back. The network you attached above have conv layers in first few layers. What I am looking is ConvX or ConvXB in all layers except first and last. So it will not be suitable for my problem. 

Lets get back to my training progress. After training darknet19-xnor for two weeks (accuracy was 67%), I attached region layer and built the following network. 
![capture](https://user-images.githubusercontent.com/24696403/51451449-adc28c00-1d78-11e9-865a-ac528c3d6c37.JPG)
Now training with voc dataset below is the training curve with mAP values. mAP improved as compared to tiny-yolov2-xnor network (25%), but still not that good! 
![darknet19-xnor-chart](https://user-images.githubusercontent.com/24696403/51451473-d64a8600-1d78-11e9-8070-3472ad661e25.png)

I will update progress. 

In addition, as I am trying to implement this paper [[A Lightweight YOLOv2: A Binarized CNN with A Parallel Support Vector Regression for an FPGA](https://dl.acm.org/citation.cfm?id=3174266)], I am struggling to prepare proper target values for SVR training. If we replace region layer with svr to predict 125 outputs in each cell, do you have any idea how we have to prepare training outputs (tx,ty,tw,th,conf,prob(20)) for SVRs training. Any suggestion will be greatly appreciated. Thank you.

**Comment by AlexeyAB** [comment]
@ghimiredhikura Hi,

> Lets get back to my training progress. After training darknet19-xnor for two weeks (accuracy was 67%), I attached region layer and built the following network.
> Now training with voc dataset below is the training curve with mAP values. mAP improved as compared to tiny-yolov2-xnor network (25%), but still not that good!

* Did you train classifier `darknet19-xnor.cfg` on ImageNet dataset? 
* How many images did you use?
* Can you rename `darknet19-xnor.cfg` to `darknet19-xnor.txt` and attach to your message?

> I am struggling to prepare proper target values for SVR training. If we replace region layer with svr to predict 125 outputs in each cell, do you have any idea how we have to prepare training outputs (tx,ty,tw,th,conf,prob(20)) for SVRs training. 

I will think how to do it in the best way.

**Comment by ghimiredhikura** [comment]
@AlexeyAB Hi,

Here is darknet19-xnor cfg file that I used. 
[darknet19-XNOR.cfg.txt](https://github.com/AlexeyAB/darknet/files/2795461/darknet19-XNOR.cfg.txt)

I used ILSVRC2012 dataset for training. 

Yes, please let me know how it will be the best way to replace REGION layer with SVR. 

Best,
Deepak

**Comment by ghimiredhikura** [comment]
@AlexeyAB Hello, 

Have you given any thoughts on how to replace region layer with SVR to predict box position and object class using linear SVRs? Any suggestions will be greatly appreciated. 

Thanks.

**Comment by Sudhakar17** [comment]
@clydebailey @AlexeyAB 
https://github.com/AlexeyAB/yolo2_light/blob/master/bin/tiny-yolo-obj_xnor.cfg
I used this config file to train two class model.
dataset : 1000 images for each class
XNOR model : 3% mAP
BW : 24%
FP: 27% 

results after 2k iterations. all models are 61MB. How to reduce the weights file size of each model?

**Comment by francis2tm** [comment]
> 
> 
> @ghimiredhikura
> 
> If I understand correctly, they try to use network like this: [yolov2_xnor_fpga.cfg.txt](https://github.com/AlexeyAB/darknet/files/2694483/yolov2_xnor_fpga.cfg.txt) - try to train this model.
> 
> This is hybrid of yolov2.cfg and yolov2-tiny.cfg. With only 1st and two last float layers, other XNOR layers.
> 
> ![image](https://user-images.githubusercontent.com/4096485/50215704-e3004680-0394-11e9-9276-15c652688f39.png)
> 
> Also check that you uncommented `difficult` in the `voc.data` file before doing `./darknet detector map`...:
> 
> ```
> classes= 20
> train  = data/train_voc.txt
> valid  = data/2007_test.txt
> difficult = data/difficult_2007_test.txt
> names = data/voc.names
> backup = backup/
> ```

Hey,
@AlexeyAB Do you know any yolo xnor-net implementation in a FPGA? With open-source software

Thanks!
